codec/abi.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package codec
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	reflect "reflect"
+)
+
+type ABIField struct {
+	Name string `json:"name"`
+	Type string `json:"type"`
+	Mask string `json:"mask,omitempty"`
+}
+
+type SingleActionABI struct {
+	ID    uint8                 `json:"id"`
+	Name  string                `json:"name"`
+	Types map[string][]ABIField `json:"types"`
+}
+
+func GetVMABIString(actions []Typed) (string, error) {
+	vmABI := make([]SingleActionABI, 0)
+	for _, action := range actions {
+		actionABI, err := getActionABI(action)
+		if err != nil {
+			return "", err
+		}
+		vmABI = append(vmABI, actionABI)
+	}
+	resBytes, err := json.MarshalIndent(vmABI, "", "  ")
+	return string(resBytes), err
+}
+
+func getActionABI(action Typed) (SingleActionABI, error) {
+	t := reflect.TypeOf(action)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+
+	result := SingleActionABI{
+		ID:    action.GetTypeID(),
+		Name:  t.Name(),
+		Types: make(map[string][]ABIField),
+	}
+
+	typesleft := []reflect.Type{t}
+	typesAlreadyProcessed := make(map[reflect.Type]bool)
```

Comment:
We should use either AvalancheGo's `set.Set` type or `map[type]struct{}` as a set